mixer/pkg/server/interceptor.go

```diff
@@ -0,0 +1,129 @@
+	// blindly lowercase the key (which is guaranteed to work in the
+	// Inject/Extract sense per the OpenTracing spec).
+	key = strings.ToLower(key)
+	w.MD[key] = append(w.MD[key], val)
+}
+
+func (w metadataReaderWriter) ForeachKey(handler func(key, val string) error) error {
+	for k, vals := range w.MD {
+		for _, v := range vals {
+			if err := handler(k, v); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+type noopSpan struct{}
+type noopSpanContext struct{}
+
+var (
+	defaultNoopSpanContext = noopSpanContext{}
+	defaultNoopSpan        = noopSpan{}
+	defaultNoopTracer      = opentracing.NoopTracer{}
+)
+
+const (
+	emptyString = ""
+)
+
+// noopSpanContext:
+func (n noopSpanContext) ForeachBaggageItem(handler func(k, v string) bool) {}
+
+// noopSpan:
+func (n noopSpan) Context() opentracing.SpanContext                       { return defaultNoopSpanContext }
+func (n noopSpan) SetBaggageItem(key, val string) opentracing.Span        { return defaultNoopSpan }
+func (n noopSpan) BaggageItem(key string) string                          { return emptyString }
+func (n noopSpan) SetTag(key string, value interface{}) opentracing.Span  { return n }
+func (n noopSpan) LogFields(fields ...log.Field)                          {}
+func (n noopSpan) LogKV(keyVals ...interface{})                           {}
+func (n noopSpan) Finish()                                                {}
+func (n noopSpan) FinishWithOptions(opts opentracing.FinishOptions)       {}
+func (n noopSpan) SetOperationName(operationName string) opentracing.Span { return n }
+func (n noopSpan) Tracer() opentracing.Tracer                             { return defaultNoopTracer }
+func (n noopSpan) LogEvent(event string)                                  {}
+func (n noopSpan) LogEventWithPayload(event string, payload interface{})  {}
+func (n noopSpan) Log(data opentracing.LogData)                           {}
+
+func TracingServerInterceptor(tracer opentracing.Tracer) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (resp interface{}, err error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			md = metadata.New(nil)
+		}
+
+		sampled := true
+		for _, val := range md.Get("x-b3-sampled") {
+			if strings.EqualFold(strings.ToLower(val), "0") {
+				sampled = false
+				break
+			}
+		}
+
+		spanContext, err := tracer.Extract(opentracing.HTTPHeaders, metadataReaderWriter{md})
+		if err != nil && err != opentracing.ErrSpanContextNotFound {
+			// TODO: establish some sort of error reporting mechanism here. We
+			// don't know where to put such an error and must rely on Tracer
+			// implementations to do something appropriate for the time being.
+
+			fmt.Printf("non nil error in trace extraction: %#v\n", err)
+		}
+		var serverSpan opentracing.Span
+
+		if !sampled || err == opentracing.ErrSpanContextNotFound {
```

Comment:
Just wondering whether it would be better to use this condition to select an appropriate tracer (e.g. real vs noop) and then just had a single StartSpan call on that selected trscer. Would save having the `noopSpan` and `noopSpanContext` duplicated here.

Reply:
maybe. this is basically throwaway quality at this point. there are probably a few better approaches; just wanted to help get an approach tested before I disappear for the holidays.